lib/Service/CategoriesService.php

```diff
@@ -0,0 +1,159 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * Calendar App
+ *
+ * @copyright 2023 Claus-Justus Heine <himself@claus-justus-heine.de>
+ *
+ * @author Claus-Justus Heine <himself@claus-justus-heine.de>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU AFFERO GENERAL PUBLIC LICENSE
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU AFFERO GENERAL PUBLIC LICENSE for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Calendar\Service;
+
+use OCP\Calendar\ICalendar;
+use OCP\Calendar\IManager as ICalendarManager;
+use OCP\IL10N;
+use OCP\SystemTag\ISystemTag;
+use OCP\SystemTag\ISystemTagManager;
+use Psr\Log\LoggerInterface;
+
+class CategoriesService {
+	/** @var null|string */
+	private $userId;
+
+	/** @var ICalendarManager */
+	private $calendarManager;
+
+	/** @var ISystemTagManager */
+	private $systemTagManager;
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	/** @var IL10N */
+	private $l;
+
+	private const CALENDAR_OBJECT_PROPERTIES_TABLE = 'calendarobjects_props';
+
+	public function __construct(?string $userId,
+		ICalendarManager $calendarManager,
+		ISystemTagManager $systemTagManager,
+		LoggerInterface $logger,
+		IL10N $l10n) {
+		$this->userId = $userId;
+		$this->calendarManager = $calendarManager;
+		$this->systemTagManager = $systemTagManager;
+		$this->logger = $logger;
+		$this->l = $l10n;
+	}
+
+	/**
+	 * This is a simplistic brute-force extraction of all already used
+	 * categories from all events accessible to the currently logged in user.
+	 *
+	 * @return array
+	 */
+	private function getUsedCategories(): array {
+		if (empty($this->userId)) {
+			return [];
+		}
+		$calendars = $this->calendarManager->getCalendarsForPrincipal('principals/users/' . $this->userId);
+		$count = count($calendars);
+		if ($count === 0) {
+			return [];
+		}
+
+		$categories = [];
+		$calendarObjects = $this->calendarManager->search('');
```

Comment:
You should be able to limit the search to events that have set a category by using
```suggestion
		$calendarObjects = $this->calendarManager->search('', ['CATEGORIES']);
```

See https://github.com/nextcloud/server/blob/be59c4efd6caf4b1e80902441287a570718b5383/apps/dav/lib/CalDAV/CalendarImpl.php#L99

Reply:
Your are right, that should work, just examined CalDavBackend.php. Late at night now, this is left for tomorrow.

Many thanks for the  hint!